includes/payment-methods/class-wc-stripe-express-checkout-custom-fields.php

```diff
@@ -0,0 +1,275 @@
+		 *
+		 * @param array $custom_checkout_data The custom checkout data.
+		 * @param WP_Error $errors The WP_Error object, for adding errors when validation fails.
+		 */
+		do_action( 'wc_stripe_express_checkout_after_checkout_validation', $custom_checkout_data, $errors );
+
+		if ( $errors->has_errors() ) {
+			$error_messages = implode( "\n", $errors->get_error_messages() );
+			throw new RouteException( 'wc_stripe_express_checkout_invalid_data', $error_messages, 400 );
+		}
+
+		/**
+		 * Allow third-party plugins to add custom checkout data for express checkout orders.
+		 *
+		 * To be used as a stand-in for the `woocommerce_checkout_update_order_meta` action.
+		 *
+		 * @since 9.6.0
+		 *
+		 * @param integer $order_id The order ID.
+		 * @param array $custom_checkout_data The custom checkout data.
+		 */
+		do_action( 'wc_stripe_express_checkout_update_order_meta', $order->get_id(), $custom_checkout_data );
+	}
+
+	/**
+	 * Get custom checkout data from the request object.
+	 *
+	 * To support custom fields in express checkout and classic checkout,
+	 * we pass custom data as extended data, i.e. under extensions.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @return array Custom checkout data.
+	 */
+	private function get_custom_checkout_data_from_request( $request ) {
+		$extensions = $request->get_param( 'extensions' );
+		if ( empty( $extensions ) ) {
+			return [];
+		}
+
+		$custom_checkout_data_json = $extensions['wc-stripe/express-checkout']['custom_checkout_data'] ?? '';
+		if ( empty( $custom_checkout_data_json ) ) {
+			return [];
+		}
+
+		$custom_checkout_data = json_decode( $custom_checkout_data_json, true );
+		if ( empty( $custom_checkout_data ) || ! is_array( $custom_checkout_data ) ) {
+			return [];
+		}
+
+		// Perform basic sanitization before passing to actions.
+		$sanitized_custom_checkout_data = [];
+		$custom_checkout_fields         = $this->get_custom_checkout_fields( 'classic' );
+		foreach ( $custom_checkout_data as $key => $value ) {
+			$field_type                                       = $custom_checkout_fields[ $key ]['type'] ?? 'text';
+			$sanitized_key                                    = sanitize_text_field( $key );
+			$sanitized_value                                  = $this->get_sanitized_field_value( $value, $field_type );
+			$sanitized_custom_checkout_data[ $sanitized_key ] = $sanitized_value;
+		}
+
+		return $sanitized_custom_checkout_data;
+	}
+
+	/**
+	 * Perform basic sanitization on custom checkout field values, based on the field type.
+	 *
+	 * @param string $value The value to sanitize.
+	 * @param string $type The type of the field.
+	 * @return mixed The sanitized field value.
+	 */
+	private function get_sanitized_field_value( $value, $type ) {
+		if ( '' === $value ) {
+			return '';
+		}
+
+		switch ( $type ) {
+			case 'checkbox':
+				return empty( $value ) ? $value : 1;
+			case 'multiselect':
+				return implode( ', ', wc_clean( $value ) );
```

Comment:
This converts the data structure from an array to a comma-separated string. Is that necessary, especially since we are inside a sanitization function?